node/subsystem/src/jaeger.rs

```diff
@@ -0,0 +1,268 @@
+	INSTANCE.read_recursive().span(hash, span_name).into()
+}
+
+/// Stateful convenience wrapper around [`mick_jaeger`].
+pub enum Jaeger {
+	/// Launched and operational state.
+	Launched {
+		/// [`mick_jaeger`] provided API to record spans to.
+		traces_in: Arc<mick_jaeger::TracesIn>,
+	},
+	/// Preparation state with the necessary config to launch the collector.
+	Prep(JaegerConfig),
+	/// Uninitialized, suggests wrong API usage if encountered.
+	None,
+}
+
+impl Jaeger {
+	/// Spawn the jaeger instance.
+	pub fn new(cfg: JaegerConfig) -> Self {
+		Jaeger::Prep(cfg)
+	}
+
+	/// Spawn the background task in order to send the tracing information out via udp
+	#[cfg(target_os = "unknown")]
+	pub fn launch(self) -> result::Result<(), JaegerError> {
+		Ok(())
+	}
+
+	/// Spawn the background task in order to send the tracing information out via udp
+	#[cfg(not(target_os = "unknown"))]
+	pub fn launch(self) -> result::Result<(), JaegerError> {
+		let cfg = match self {
+			Self::Prep(cfg) => Ok(cfg),
+			Self::Launched{ .. } => {
+				return Err(JaegerError::AlreadyLaunched)
+			}
+			Self::None => Err(JaegerError::MissingConfiguration),
+		}?;
+
+		let jaeger_agent = cfg.agent_addr;
+
+		log::info!("🐹 Collecting jaeger spans for {:?}", &jaeger_agent);
+
+		let (traces_in, mut traces_out) = mick_jaeger::init(mick_jaeger::Config {
+			service_name: format!("{}-{}", cfg.node_name, cfg.node_name),
+		});
+
+
+		// Spawn a background task that pulls span information and sends them on the network.
+		let _handle = async_std::task::spawn::<_, result::Result<(), JaegerError>>(async move {
+			let mut port = 49000_u16;
+			let mut udp_socket;
+
+			loop {
+				udp_socket = async_std::net::UdpSocket::bind(format!("127.0.0.1:{}", port)).await;
+				if udp_socket.is_ok() {
+					break;
+				}
+				port += 1;
+				if port == std::primitive::u16::MAX {
+					break;
+				}
+			}
+			let udp_socket = udp_socket.map_err(|e| JaegerError::PortAllocationError(e))?;
+
+			loop {
+				let buf = traces_out.next().await;
+				// UDP sending errors happen only either if the API is misused or in case of missing privilege.
+				if let Err(e) = udp_socket.send_to(&buf, jaeger_agent).await
+					.map_err(|e| JaegerError::SendError(e))
+				{
+					log::trace!("Failed to send jaeger span: {:?}", e);
+				}
+			}
+		});
+
+		*INSTANCE.write() = Self::Launched {
+			traces_in,
+		};
+		Ok(())
+	}
+
+	#[inline(always)]
+	fn traces_in(&self) -> Option<&Arc<mick_jaeger::TracesIn>> {
+		match self {
+			Self::Launched {
+				traces_in,
+				..
+			} => Some(&traces_in),
+			_ => None,
+		}
+	}
+
+	fn span(&self, hash: &Hash, span_name: impl Into<String>) -> Option<mick_jaeger::Span> {
+		if let Some(traces_in) = self.traces_in() {
+			let trace_id = {
+				let mut buf = [0u8; 16];
+				buf.copy_from_slice(&hash.as_ref()[0..16]);
+				std::num::NonZeroU128::new(u128::from_be_bytes(buf))
+			}.expect("NonZeroU128::new returns `Some(_)` for 16 bytes long slice; qed");
```

Comment:
sorry if I wasn't clear, this will panic if the provided `Hash` is zero and it's not clear to me why it can't be (accidentally)

Reply:
`NonZeroU128` it already says non-zer0, I need a :coffee:

Reply:
I mean, it's possible it would accidentally hash to 0, but at 1 trillion hashes per second, it would still take about 1000 times the age of the universe (if my math is right) to accidentally compute it.

Reply:
We already had a bug when chain-api accidentally returned a zero hash: https://github.com/paritytech/polkadot/pull/1795/commits/ba768b4155ece78f35741e0f1ac014c5138c7bdc. This was far from obvious and I'm not sure about bringing down the node on such an error.

Reply:
In that case it was returning a default, zeroized hash in one case. Here, it's calculating a new hash; my point is that a new hash calculation is _very_ unlikely to return 128 0 bits (assuming the hash algorithm itself is appropriate).

Reply:
It's not calculating any hashes, it just copies the provided one, or am I missing something?

Reply:
No, I missed something: you're right. There's nothing currently stopping this from panicing on an invalid hash, and we aren't currently validating that. 

Probably the simplest solution is to return `None` on an invalid input hash.

Reply:
I agree, returning `None` here is sensible.

Reply:
Actually, the provided hash can be perfectly valid with the first 16 bytes of zero (highly unlikely though).
We only copy the first 16 bytes, while the actual hash is 32 bytes long.

Reply:
Right, which is why we can propagate the output of the `NonZeroU128` constructor if `None`.

Reply:
I agree, this should return `None` in those cases